### PR TITLE
RUN: Support test window

### PR DESCRIPTION
**Pull request by mchernyavsky**
Closes https://github.com/intellij-rust/intellij-rust/issues/1485.

This PR is mostly based on https://github.com/intellij-rust/intellij-rust/pull/2337 (a huge thanks to @farodin91 and @mkaput).

screen shot 2018-10-09 at 22 13 58

I had to remove the `--nocapture` option because of https://github.com/rust-lang/rust/issues/54669.

TODO:
- [x] Provide execution time of each test
- [x] Provide location hint and proper line marker for test suites
- [ ] Add more tests

What else can be implemented/improved?
- [ ] Add more filters
- [ ] Support "Run only tests that failed/crashed after last run"

### CODE REVIEW

File: src/main/kotlin/org/rust/cargo/runconfig/test/CargoTestEventsConverter.kt [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+
+        return false
+    }
+
+    /** @return true if message successfully processed. */
+    private fun handleSimpleMessage(text: String, visitor: ServiceMessageVisitor): Boolean {
+        fun startsWith(prefix: String): Boolean = text.trim().startsWith(prefix, true)
+        return when {
+            startsWith("compiling") -> {
+                visitor.visitCompilationStarted(CompilationStarted("rustc"))
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/cargo/runconfig/test/CargoTestEventsConverter.kt]: Extract magic strings

File: src/main/kotlin/org/rust/cargo/runconfig/test/CargoTestEventsConverter.kt [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+import com.intellij.execution.testframework.sm.runner.OutputToGeneralTestEventsConverter
+import com.intellij.openapi.util.Key
+import jetbrains.buildServer.messages.serviceMessages.CompilationFinished
+import jetbrains.buildServer.messages.serviceMessages.CompilationStarted
+import jetbrains.buildServer.messages.serviceMessages.ServiceMessageVisitor
+import org.rust.cargo.runconfig.test.CargoTestLocator.getTestFnUrl
+import org.rust.cargo.runconfig.test.CargoTestLocator.getTestModUrl
+import org.rust.stdext.removeLast
+
+typealias NodeId = String
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/cargo/runconfig/test/CargoTestEventsConverter.kt]: Can be private?

File: src/main/kotlin/org/rust/cargo/runconfig/test/CargoTestEventsConverter.kt [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+    private fun createServiceMessagesFor(suiteMessage: LibtestSuiteMessage): List<ServiceMessageBuilder>? {
+        val messages = mutableListOf<ServiceMessageBuilder>()
+        when (suiteMessage.event) {
+            "started" -> {
+                if (suiteMessage.test_count.toInt() == 0) return emptyList()
+                messages.add(testSuiteStarted(target))
+                messages.add(testCount(suiteMessage.test_count))
+            }
+            "ok", "failed" -> {
+                pendingFinishedSuites.mapTo(messages) { testSuiteFinished(it) }
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/cargo/runconfig/test/CargoTestEventsConverter.kt]: Double-check that new suite cannot be started before previous is finished

File: src/main/kotlin/org/rust/cargo/runconfig/test/CargoTestEventsConverter.kt [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+        suitesStack.add(suite)
+    }
+
+    companion object {
+        private val NodeId.name: String
+            get() = substringAfterLast(".")
+        private val NodeId.parent: NodeId
+            get() {
+                val parent = substringBeforeLast(".")
+                return if (this == parent) "0" else parent
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/cargo/runconfig/test/CargoTestEventsConverter.kt]: Let's extract "0" to constant named like "something_ROOT" :)

File: src/main/kotlin/org/rust/cargo/runconfig/test/CargoTestEventsConverter.kt [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+            }
+
+        private fun testSuiteStarted(suite: NodeId): ServiceMessageBuilder =
+            ServiceMessageBuilder.testSuiteStarted(suite.name)
+                .addAttribute("nodeId", suite)
+                .addAttribute("parentNodeId", suite.parent)
+                .addAttribute("locationHint", getTestModUrl(suite.replace(".", "::")))
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/cargo/runconfig/test/CargoTestEventsConverter.kt]: Let's create extension property for `.replace(".", "::")`

File: src/main/kotlin/org/rust/cargo/runconfig/test/CargoTestEventsConverter.kt [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+        private fun testFinished(test: NodeId, duration: String): ServiceMessageBuilder =
+            ServiceMessageBuilder.testFinished(test.name)
+                .addAttribute("nodeId", test)
+                .addAttribute("duration", duration)
+
+        private fun testIgnored(test: NodeId): ServiceMessageBuilder =
+            ServiceMessageBuilder.testIgnored(test.name)
+                .addAttribute("nodeId", test)
+
+        private fun testCount(testCount: NodeId): ServiceMessageBuilder =
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/cargo/runconfig/test/CargoTestEventsConverter.kt]: NodeId -> String

**Comment by mchernyavsky** [comment]
bors retry

**Comment by mchernyavsky** [comment]
bors retry

**Comment by mchernyavsky** [comment]
bors r=vlad20012

**Pull request merged**